### PR TITLE
Fix: handle branch names mismatch

### PR DESCRIPTION
**Pull request by vanessayuenn**
### Description of the Change

This PR introduces the use of `refspec` in push and pull operations (fetch does it a bit differently), in order to handle situations where the local branch is named differently from its tracking branch. This set of circumstances has become less of an edge case since the introduction of PR-checkout -- we prefix the branch name with PR number, etc. 

### Benefits

- Able to fetch, pull, push checked out PRs
- Able to be creative in naming your local branches 👯 

### Possible Drawbacks

This PR changes how the underlying git operations work, which is used by many things in the package; so the possible setback is if things break in edge cases that I haven't thought of.

### Applicable Issues

fixes: #1697 
fixes: #918 

### Metrics

N/A

### Tests

Added unit tests to verify that push & pull & fetch work as expected despite the difference between local branch name and its upstream branch name. 

Manually tested push & pull & fetch operations in these cases:
- local branch named differently from upstream
- local branch named the same as upstream
- checked out PR from a branch in the original repo
- checked out PR opened from a forked repo

### CODE REVIEW

File: lib/models/branch.js [hunk 1]
```diff
@@ -67,6 +67,23 @@ export default class Branch {
...
+    }
+    const remoteBranch = action === 'PUSH' ? this.push : this.upstream;
+    const remoteBranchName = remoteBranch.getShortRemoteRef();
+    const localBranchName = this.getName();
+    if (remoteBranchName && remoteBranchName !== localBranchName) {
+      if (action === 'PUSH') {
+        return `${localBranchName}:${remoteBranchName}`;
+      } else if (action === 'PULL') {
+        return `${remoteBranchName}:${localBranchName}`;
+      }
```
Comment (smashwilson) [inline lib/models/branch.js]: Using strings for enumerated values always makes me a bit nervous - it feels like it's easy to typo them somewhere (or miss a callsite when refactoring, say) and get surprising behavior. Sometimes I'll use Symbols instead, or objects with a `.when()` method, to push the "getting a valid action" work down into the language more.

With that said: the real answer here is "port our package to TypeScript" anyway. So this is just fine for now :smile:

File: test/controllers/status-bar-tile-controller.test.js [hunk 2]
```diff
@@ -646,6 +646,50 @@ describe('StatusBarTileController', function() {
...
+
+    it('pulls from the correct branch', async function() {
+      const prePullSHA = await repository.git.exec(['rev-parse', 'HEAD']);
+      await repository.git.exec(['reset', '--hard', 'HEAD~2']);
+      sinon.spy(repository, 'pull');
+      await wrapper
+        .instance()
+        .pull(await wrapper.instance().fetchData(repository))();
+      const postPullSHA = await repository.git.exec(['rev-parse', 'HEAD']);
+      assert.isTrue(repository.pull.called);
```
Comment (annthurium) [inline test/controllers/status-bar-tile-controller.test.js]: nit: do you want to assert that `repository.pull` was called with the args you expect?

File: test/controllers/status-bar-tile-controller.test.js [hunk 3]
```diff
@@ -646,6 +646,50 @@ describe('StatusBarTileController', function() {
...
+
+    it('pushes to the correct branch', async function() {
+      await repository.git.commit('new local commit', {allowEmpty: true});
+      const localSHA = await repository.git.exec(['rev-parse', 'another-name']);
+      sinon.spy(repository, 'push');
+      await wrapper
+        .instance()
+        .push(await wrapper.instance().fetchData(repository))();
+      const remoteSHA = await repository.git.exec(['rev-parse', 'origin/master']);
+      assert.isTrue(repository.push.called);
```
Comment (annthurium) [inline test/controllers/status-bar-tile-controller.test.js]: same nit as above - might be good to specify which args `push` should be called with.
